students/terrance_jones/lesson5/mailroom_part3.py

```diff
@@ -0,0 +1,169 @@
+            amount = float(amount)
+        except ValueError:
+            print("Donation amount must be a numeric value")
+        
+        else:
+            return amount
+
+def add_to_existing_donor(d, a):
+    """adds donation amount to donor in list. d is donor name. a is amount of donation"""
+    donor_list[d].append(a)
+
+def create_new_donor(d, a):
+    donor_list[d] = [a]
+
+def print_rows():
+    for c in donor_list:
+        name = c
+        num_gifts = len(donor_list[c])
+        total = sum(donor_list[c])
+        average = total / num_gifts
+        average = round(average,2)
+
+        
+        
+        print("{:<25} ${:^20}  {:^15} ${:>20}".format(name,total,num_gifts,average))
+
+
+def letter(donor):
+    output = {'name': donor}
+    all_donations = donor_list[donor]
+    output = {'name': donor, 'total': sum(donor_list[donor])}
+   
+    output['last_donation']= all_donations[len(all_donations)-1]
+    letter = '''Dear {name},\n
+    \tThank you for your most recent donation of ${last_donation:,.2f}
+    \tYour total support of {total:,.2f} has been used to assist many people.\n
+    Thank you,\n
+    -The Mailroom'''.format(**output)
+
+    return(letter)
+
+
+
+#thank you function gets a donors name and amount of donation. then prints out a thank you email
+def thankyou():
+    donor_name = get_name()
+    if donor_name == "list":
+        show_donor_list
+        donor_name =get_name()
+
+    if find_donor(donor_name) is True: 
+        donation_amount = get_amount()
+        add_to_existing_donor(donor_name, donation_amount)
+        
+        dic = {'name': donor_name , 'amount_donated': donation_amount}
+        print("Dear {name}, \n Thank you for your {amount_donated:,.2f} donation. \nSincerely, Mailroom.".format(**dic))
+        
+    else:
+        
+        donation_amount = get_amount()
+        create_new_donor(donor_name, donation_amount)
+        dic = {'name': donor_name , 'amount_donated': donation_amount}
+        print("Dear {name}, \n Thank you for your {amount_donated:,.2f} donation. \nSincerely, Mailroom.".format(**dic))
+       
+            
+
+
+def create_report():
+    
+    heading = "{:<25s} | {:^20s} | {:^10s} | {:^20s} ".format("Donor Name", "Total Given", "Num Gifts", "Average Gift")
+    print(heading)
+    print("-" * len(heading))
+    print_rows()
+   
+
+def letters_everyone():
+    for key in donor_list.keys():
+        filename = "{}.txt".format(key)
+        with open(filename, 'w') as f:
+            donor_letter = letter(key)
+            f.write(donor_letter)
+    print("Letters have been created for each donor")
+   
+
+def menu_selection(prompt, dispatch_dict):
+    """displays the menu. asks for user to select option. if option is not available will throw KeyError and ask the user for new input"""
+    while True:
+        try:
+            response = input(prompt)
+            if dispatch_dict[response]() == "exit menu":
+                return
+        except KeyError:
```

Comment:
nice - it is best to put the exception you're catching only around the code that can throw it (which would wrap around lines 119-120)